File: java/src/main/kthSmallestEleInBST/recursive.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int left = totalNodes(root.left) ;
        if(k == left +1){
            return root.val;
        }
        else if( k <= left ){
            return kthSmallest(root.left, k);
        }
        else{
            return kthSmallest(root.right, k - left-1);
        }
    }
    
    public int totalNodes(TreeNode root){
        if(root == null){
            return 0;
        }
        return totalNodes(root.left) + totalNodes(root.right) + 1;
    }
}

File: python/99RecoverBinarySearchTree/recursive.py
class TreeNode:
    def __init__(self, val=0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def recoverTree(self, root = Optional[TreeNode]) -> None:
        node = []
        def append_function(seq, node):
            seq.append(node.val)
        def pop_function(seq, node):
            node.val = seq.pop(0)

        def inorder_traversal(root, func):
            if not root:
                return
            inorder_traversal(root.left, func)
            func(node, root)
            inorder_traversal(root.right, func)
        inorder_traversal(root, append_func)
        node.sort()
        inorder_traversal(root, pop_func)

